### PR TITLE
Nsaxena/Create Account : Take out invitation key

### PR DESCRIPTION
**Pull request by NehaBroad**

### CODE REVIEW

File: ui/src/app/app.module.ts [hunk 1]
```diff
@@ -76,7 +76,8 @@ export function getConfiguration(signInService: SignInService): Configuration {
     ReviewComponent,
     WorkspaceComponent,
     WorkspaceEditComponent,
-    WorkspaceShareComponent
+    WorkspaceShareComponent,
+      InvitationCodeComponent
```
Comment (blrubenstein) [inline ui/src/app/app.module.ts]: Nit: Spacing

File: ui/src/app/views/app/component.html [hunk 2]
```diff
@@ -41,7 +41,8 @@
         </div>
       </div>
       <div *ngIf="_showCreateAccount;">
-        <app-account-creation></app-account-creation>
+        <!--<app-account-creation></app-account-creation>-->
+          <app-invitation-code></app-invitation-code>
```
Comment (blrubenstein) [inline ui/src/app/views/app/component.html]: Do we need the commented `<app-account-creation>` element anymore?

File: ui/src/app/views/invitation-code/component.spec.ts [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    it('required invitation code', fakeAsync(() => {
+
+        simulateClick(invitationCodePage.fixture, invitationCodePage.nextButton);
+        const app = invitationCodePage.fixture.debugElement.componentInstance;
+
+        expect(app.invitationSucc).toBeFalsy();
+        expect(app.invitationKeyReq).toBeTruthy();
+        expect(app.invitationKeyInvalid).toBeFalsy();
+    }));
+    it('invalid invitation code', fakeAsync(() => {
```
Comment (blrubenstein) [inline ui/src/app/views/invitation-code/component.spec.ts]: For test names, I've been trying to make it a story based on what I am testing. For example, for this, I would recommend something like:
`it('does not continue with an invalid invitation code',...)`

That said, we haven't talked a ton about test naming conventions yet, so that may be a good conversation to have.

File: ui/src/testing/stubs/profile-service-stub.ts [hunk 4]
```diff
@@ -12,7 +13,8 @@ export class ProfileStubVariables {
     fullName:  'Tester MacTesterson><script>alert("hello");</script>',
     givenName: 'Tester!@#$%^&*()><script>alert("hello");</script>',
     familyName: 'MacTesterson!@#$%^&*()><script>alert("hello");</script>',
-    phoneNumber: '999-999-9999'
+    phoneNumber: '999-999-9999',
+      invitationKey : 'dummyKey'
```
Comment (blrubenstein) [inline ui/src/testing/stubs/profile-service-stub.ts]: Nit: Spacing

File: api/src/main/java/org/pmiops/workbench/api/ProfileController.java [hunk 5]
```diff
@@ -360,4 +357,14 @@ private User initializeUserIfNeeded() {
 
     return ResponseEntity.status(HttpStatus.NO_CONTENT).build();
   }
+
+  @Override
+  public ResponseEntity<Void> invitationCodeVerification(InvitationVerRequest invitationVerRequest){
+    if (invitationVerRequest.getInvitationKey() == null
+            || !invitationVerRequest.getInvitationKey().equals(cloudStorageService.readInvitationKey())) {
```
Comment (blrubenstein) [inline api/src/main/java/org/pmiops/workbench/api/ProfileController.java]: Nit: Spacing

File: api/src/main/resources/workbench.yaml [hunk 6]
```diff
@@ -1839,9 +1860,14 @@ definitions:
         $ref: '#/definitions/Profile'
       password:
         type: string
-      invitationKey:
-        type: string
-        description: Invitation key. Necessary when creating account before public launch.
+
+  InvitationVerRequest:
+      type: object
```
Comment (blrubenstein) [inline api/src/main/resources/workbench.yaml]: Nit: Spacing

File: ui/src/app/views/app/component.html [hunk 7]
```diff
@@ -41,7 +41,7 @@
         </div>
       </div>
       <div *ngIf="_showCreateAccount;">
-        <app-account-creation></app-account-creation>
+          <app-invitation-code></app-invitation-code>
```
Comment (blrubenstein) [inline ui/src/app/views/app/component.html]: Nit: Spacing

File: ui/src/app/views/invitation-code/component.css [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+}
+
+.error {
+  background-color: #ffeeee;
+  padding: 0 0.5rem;
+  margin-top: 0.2rem;
+}
+
+.label{
+  width:10px;
```
Comment (blrubenstein) [inline ui/src/app/views/invitation-code/component.css]: Is there a specific reason to use `px` over `rem` here?

File: ui/src/app/views/invitation-code/component.html [hunk 9]
```diff
@@ -0,0 +1,23 @@
+<div style="padding: 5rem;">
+    <div *ngIf="!invitationSucc" class="form-area">
+            <div class="form-section">
```
Comment (blrubenstein) [inline ui/src/app/views/invitation-code/component.html]: Nit: Spacing

File: ui/src/app/views/invitation-code/component.ts [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+})
+
+export class InvitationCodeComponent {
+
+    invitationKey: string;
+    invitationSucc: boolean;
+    invitationKeyReq: boolean;
+    invitationKeyInvalid: boolean;
+    constructor(private profileService: ProfileService) {
+        this.invitationSucc = false;
```
Comment (blrubenstein) [inline ui/src/app/views/invitation-code/component.ts]: Nit: Spacing

File: ui/src/testing/stubs/profile-service-stub.ts [hunk 11]
```diff
@@ -31,4 +33,19 @@ export class ProfileServiceStub {
       }, 0);
     });
   }
+
+  public invitationCodeVerification(invitationVerRequest?: InvitationVerRequest): Observable<{}> {
+          if (invitationVerRequest.invitationKey === 'dummy') {
```
Comment (blrubenstein) [inline ui/src/testing/stubs/profile-service-stub.ts]: Nit: Spacing

File: ui/src/app/views/invitation-code/component.css (hunk 8 above)
Comment (NehaBroad) [reply ui/src/app/views/invitation-code/component.css]: No particular reason...since was not sure about the styling i used px....will change it to rem

File: api/src/main/java/org/pmiops/workbench/api/ProfileController.java [hunk 12]
```diff
@@ -258,11 +259,7 @@ private User initializeUserIfNeeded() {
 
   @Override
   public ResponseEntity<Profile> createAccount(CreateAccountRequest request) {
-    if (request.getInvitationKey() == null
```
Comment (dmohs) [inline api/src/main/java/org/pmiops/workbench/api/ProfileController.java]: This key needs to be checked during account creation. Otherwise, I can create an account without an invitation key by using curl.

File: api/src/main/resources/workbench.yaml [hunk 13]
```diff
@@ -111,6 +111,27 @@ paths:
           schema:
             $ref: "#/definitions/UsernameTakenResponse"
 
+  /api/v1/google-invitation:
```
Comment (dmohs) [inline api/src/main/resources/workbench.yaml]: How is this code related to Google? In other words, what is the significance of "google" in this path?

File: api/src/main/resources/workbench.yaml [hunk 14]
```diff
@@ -1839,9 +1860,14 @@ definitions:
         $ref: '#/definitions/Profile'
       password:
         type: string
+
+  InvitationVerRequest:
```
Comment (dmohs) [inline api/src/main/resources/workbench.yaml]: I got stuck on this, so I think we need to just spell it out as `InvitationCodeVerificationRequest`.

File: ui/src/app/views/invitation-code/component.spec.ts [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+import {ErrorHandlingService} from '../../services/error-handling.service';
+import {SignInService} from '../../services/sign-in.service';
+import {AccountCreationComponent} from '../account-creation/component';
+import {InvitationCodeComponent} from '../invitation-code/component';
+
+
+
+
+
+class InvitationCodePage {
```
Comment (dmohs) [inline ui/src/app/views/invitation-code/component.spec.ts]: Let's visit this file together to work on spacing.

File: ui/src/app/views/invitation-code/component.ts [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+@Component ({
+    selector : 'app-invitation-code',
+    styleUrls: ['./component.css'],
+    templateUrl: './component.html'
+})
+
+export class InvitationCodeComponent {
+
+    invitationKey: string;
+    invitationSucc: boolean;
```
Comment (dmohs) [inline ui/src/app/views/invitation-code/component.ts]: I'm not sure you're saving much with this abbreviation.

File: ui/src/testing/stubs/profile-service-stub.ts [hunk 17]
```diff
@@ -12,7 +13,8 @@ export class ProfileStubVariables {
     fullName:  'Tester MacTesterson><script>alert("hello");</script>',
     givenName: 'Tester!@#$%^&*()><script>alert("hello");</script>',
     familyName: 'MacTesterson!@#$%^&*()><script>alert("hello");</script>',
-    phoneNumber: '999-999-9999'
+    phoneNumber: '999-999-9999',
+    invitationKey : 'dummyKey'
```
Comment (NehaBroad) [inline ui/src/testing/stubs/profile-service-stub.ts]: Nit: Spacing

File: ui/src/app/views/invitation-key/component.html [hunk 18]
```diff
@@ -0,0 +1,22 @@
+<div style="padding: 5rem;">
+    <div *ngIf="!invitationKeyVerifed" class="form-area">
+        <div class="form-section">
+        <label>Enter you Invitation Key:</label>
```
Comment (blrubenstein) [inline ui/src/app/views/invitation-key/component.html]: Nit: 'Enter your Invitation Key:'

File: ui/src/app/views/invitation-key/component.spec.ts [hunk 19]
```diff
@@ -0,0 +1,108 @@
...
+import {ProfileServiceStub} from 'testing/stubs/profile-service-stub';
+
+import {
+    queryByCss, simulateClick,
+    updateAndTick
+} from '../../../testing/test-helpers';
+import {ErrorHandlingService} from '../../services/error-handling.service';
+import {SignInService} from '../../services/sign-in.service';
+import {AccountCreationComponent} from '../account-creation/component';
+import {InvitationKeyComponent} from './/component';
```
Comment (blrubenstein) [inline ui/src/app/views/invitation-key/component.spec.ts]: `./invitation-key/component` for the import line?

File: api/src/main/resources/workbench.yaml [hunk 20]
```diff
@@ -111,6 +111,27 @@ paths:
           schema:
             $ref: "#/definitions/UsernameTakenResponse"
 
+  /api/v1/invitationKeyVerification:
```
Comment (dmohs) [inline api/src/main/resources/workbench.yaml]: Any particular reason you chose camel-case here instead of hyphen-separated?

File: api/src/main/java/org/pmiops/workbench/api/ProfileController.java [hunk 21]
```diff
@@ -362,6 +363,18 @@ private User initializeUserIfNeeded() {
   }
 
   @Override
+  public ResponseEntity<Void> invitationKeyVerification(InvitationVerRequest invitationVerRequest){
+    if (!verifyInvitationKey(invitationVerRequest.getInvitationKey())) {
+        throw new BadRequestException(
```
Comment (dmohs) [inline api/src/main/java/org/pmiops/workbench/api/ProfileController.java]: Can you eliminate the duplication in this throw?

File: api/src/main/resources/workbench.yaml [hunk 22]
```diff
@@ -111,6 +111,27 @@ paths:
...
+      parameters:
+        - in: body
+          name: invitationVerRequest
+          schema:
+            $ref: "#/definitions/InvitationVerRequest"
+      responses:
+        "200":
+          description: Invitation Key verified.
+        "400":
+          description: Error occurred while verifying Invitation Key .
```
Comment (dmohs) [inline api/src/main/resources/workbench.yaml]: Nit: spacing

File: api/src/main/resources/workbench.yaml [hunk 23]
```diff
@@ -1911,9 +1932,16 @@ definitions:
         $ref: '#/definitions/Profile'
       password:
         type: string
+      invitationKey:
+              type: string
```
Comment (dmohs) [inline api/src/main/resources/workbench.yaml]: Nit: spacing

File: api/src/main/resources/workbench.yaml [hunk 24]
```diff
@@ -1911,9 +1932,16 @@ definitions:
...
+
+  InvitationVerRequest:
+    type: object
+    required:
+      - invitationKey
+    properties:
       invitationKey:
         type: string
-        description: Invitation key. Necessary when creating account before public launch.
```
Comment (dmohs) [inline api/src/main/resources/workbench.yaml]: Why remove this description?

File: api/src/main/resources/workbench.yaml [hunk 25]
```diff
@@ -1911,9 +1932,16 @@ definitions:
         $ref: '#/definitions/Profile'
       password:
         type: string
+      invitationKey:
+              type: string
+
+  InvitationVerRequest:
```
Comment (dmohs) [inline api/src/main/resources/workbench.yaml]: Still not crazy about this shortening. Did we already discuss?

File: api/src/test/java/org/pmiops/workbench/api/ProfileControllerTest.java [hunk 26]
```diff
@@ -58,296 +54,304 @@
...
-        .when(fireCloudService).createAllOfUsBillingProject(projectName + "-3");
-    doThrow(new ApiException(HttpStatus.CONFLICT.value(), "conflict"))
-        .when(fireCloudService).createAllOfUsBillingProject(projectName + "-4");
-
-    try {
-      cloudProfileController.getMe();
-      fail("ServerErrorException expected");
-    } catch (ServerErrorException e) {
-      // expected
+    private static final Instant NOW = Instant.now();
```
Comment (dmohs) [inline api/src/test/java/org/pmiops/workbench/api/ProfileControllerTest.java]: The changes in this file are mostly a conversion from two spaces (correct) to four spaces (incorrect). Is your editor configured correctly?

File: ui/src/app/views/invitation-key/component.css [hunk 27]
```diff
@@ -0,0 +1,20 @@
...
+}
+
+.error {
+  background-color: #ffeeee;
+  padding: 0 0.5rem;
+  margin-top: 0.2rem;
+}
+
+.label{
+  width:3rem;
```
Comment (dmohs) [inline ui/src/app/views/invitation-key/component.css]: Nit: spacing.

File: ui/src/app/views/invitation-key/component.html [hunk 28]
```diff
@@ -0,0 +1,22 @@
+<div style="padding: 5rem;">
+    <div *ngIf="!invitationKeyVerifed" class="form-area">
```
Comment (dmohs) [inline ui/src/app/views/invitation-key/component.html]: Two-space indents.

File: ui/src/app/views/invitation-key/component.html [hunk 29]
```diff
@@ -0,0 +1,22 @@
...
+        <div style="display: flex; justify-content: space-around;">
+            <button id="next" (click)="next()" class="btn btn-primary">Next</button>
+        </div>
+    </div>
+    <div *ngIf="invitationKeyVerifed">
+        <app-account-creation></app-account-creation>
+    </div>
+
+
+</div>
```
Comment (dmohs) [inline ui/src/app/views/invitation-key/component.html]: Let's also configure your editor to insert a newline at the end of each file so we don't get these.

File: api/src/main/resources/workbench.yaml (hunk 20 above)
Comment (NehaBroad) [reply api/src/main/resources/workbench.yaml]: It was just a personal Choice, i prefer camelCase over hyphen, but if hyphen is the norm i can change that

File: api/src/main/resources/workbench.yaml (hunk 22 above)
Comment (NehaBroad) [reply api/src/main/resources/workbench.yaml]: We can discuss about this tomorrow, as i did check and it has two spaces, i might be missing something.

File: api/src/main/resources/workbench.yaml (hunk 24 above)
Comment (NehaBroad) [reply api/src/main/resources/workbench.yaml]: It happened by mistake have added it again

File: api/src/main/resources/workbench.yaml (hunk 20 above)
Comment (dmohs) [reply api/src/main/resources/workbench.yaml]: Since this is a URL, I have urged the team to move toward hyphen-separated words. URLs (even API URLs occasionally) tend to wind up in a user's browser address bar. We shouldn't expect users to understand case-sensitivity and, for similar reasons, I also discourage snake-case (foo_bar) because underscores are often hard to spot.

File: ui/src/app/views/invitation-key/component.spec.ts [hunk 30]
```diff
@@ -0,0 +1,102 @@
...
+      ],
+      declarations: [
+        AccountCreationComponent,
+        InvitationKeyComponent
+      ],
+      providers: [
+        { provide: ErrorHandlingService, useValue: new ErrorHandlingServiceStub() },
+        { provide: SignInService, useValue: {} },
+        { provide: ProfileService, useValue: new ProfileServiceStub() }
+      ] }).compileComponents().then(() => {
```
Comment (dmohs) [inline ui/src/app/views/invitation-key/component.spec.ts]: These next few lines catch my eye. Is the spacing right here?

File: ui/src/app/views/invitation-key/component.ts [hunk 31]
```diff
@@ -0,0 +1,47 @@
...
+  invitationKeyVerifed: boolean;
+  invitationKeyReq: boolean;
+  invitationKeyInvalid: boolean;
+  constructor(
+    private profileService: ProfileService
+  ) {
+      this.invitationKeyVerifed = false;
+      this.invitationKeyReq = false;
+      this.invitationKeyInvalid = false;
+    }
```
Comment (dmohs) [inline ui/src/app/views/invitation-key/component.ts]: Nit: spacing? (doesn't look right to me but maybe it is)

File: ui/src/testing/stubs/profile-service-stub.ts [hunk 32]
```diff
@@ -31,4 +33,18 @@ export class ProfileServiceStub {
       }, 0);
     });
   }
+
+  public invitationKeyVerification(invitationVerificationRequest?: InvitationVerificationRequest)
+      : Observable<{}> {
+    if (invitationVerificationRequest.invitationKey === 'dummy') {
+      const observable = new Observable(observer => {
+          observer.next(this.profile);
```
Comment (dmohs) [inline ui/src/testing/stubs/profile-service-stub.ts]: Nit: spacing here and downward looks a bit funny to me.

File: ui/src/app/views/invitation-key/component.spec.ts [hunk 33]
```diff
@@ -0,0 +1,102 @@
...
+    queryByCss, simulateClick,
+    updateAndTick
+} from '../../../testing/test-helpers';
+import {ErrorHandlingService} from '../../services/error-handling.service';
+import {SignInService} from '../../services/sign-in.service';
+import {AccountCreationComponent} from '../account-creation/component';
+import {InvitationKeyComponent} from '../invitation-key/component';
+
+
+
```
Comment (dmohs) [inline ui/src/app/views/invitation-key/component.spec.ts]: I don't think this pattern (lots of space after imports) exists anywhere else in our code. Let's knock out a couple of these newlines.

File: ui/src/app/views/invitation-key/component.css (hunk 27 above)
Comment (dmohs) [reply ui/src/app/views/invitation-key/component.css]: Neha, it looks like you missed a few of my comments. Is that true? If so, could you open a follow-up PR to address them?
